### PR TITLE
fix(eslint-plugin): add getConstraintInfo to handle generic constraints better

### LINKED ISSUE #10438: Repo: Investigate better handling of unconstrained generics
### Suggestion

Sadly, right now `getConstrainedTypeOfLocation` is bugged for unconstrained generics. The code just gives back the type parameter type if getBaseConstraintOfType(nodeType) is nullish, but that's not correct. Unconstrained generics should behave like `unknown`. Here's what it is right now:

https://github.com/typescript-eslint/typescript-eslint/blob/62d57559564fb08512eafe03a2c1b167c4377601/packages/type-utils/src/getConstrainedTypeAtLocation.ts#L7-L20

And here's what it "should" be: 

```ts
/**
 * Resolves the given node's type. Will resolve to the type's generic constraint, if it has one.
 */
export function getConstrainedTypeAtLocation(
  services: ParserServicesWithTypeInformation,
  node: TSESTree.Node,
): ts.Type {
  const nodeType = services.getTypeAtLocation(node);
  if (tsutils.isTypeParameter(nodeType)) {
    const checker = services.program.getTypeChecker();

    return (
      checker.getBaseConstraintOfType(nodeType) ?? checker.getUnknownType()
    );
  }
  return nodeType;
}
```

Unfortunately, due to https://github.com/microsoft/TypeScript/issues/60475, there is no `checker.getUnknownType()` yet or for a while.

It kind of sucks, but I propose that we instead modify the implementation to something more like
```ts
export function getConstrainedTypeAtLocation(
  services: ParserServicesWithTypeInformation,
  node: TSESTree.Node,
): ts.Type | undefined { // <-- note the `| undefined`
  const nodeType = services.getTypeAtLocation(node);
  if (tsutils.isTypeParameter(nodeType)) {
    const checker = services.program.getTypeChecker();
    return checker.getBaseConstraintOfType(nodeType) 
    /* in the future this can have ?? checker.getUnknownType() */;
  }
  return nodeType;
}
```
and force callers to handle the unconstrained case separately.

***

Note - I think we should also double-check (and add tests) for what happens when the constraint is `any`. I'm pretty sure `function f<T extends any>(x: T) {}` behaves as though `T` were `unknown` rather than `any`. So maybe we should really be doing

```ts
export function getConstrainedTypeAtLocation(
  services: ParserServicesWithTypeInformation,
  node: TSESTree.Node,
): ts.Type | undefined {
  const nodeType = services.getTypeAtLocation(node);
  if (tsutils.isTypeParameter(nodeType)) {
    const checker = services.program.getTypeChecker();
    const constraint = checker.getBaseConstraintOfType(nodeType);
    return constraint == null || isTypeAnyType(constraint)
      ? /* in the future this can be checker.getUnknownType() */ undefined
      : constraint;
  }
  return nodeType;
}

```

I haven't yet checked what `checker.getBaseConstraintOfType()` does with an `any`.

### Additional Info

Related, caused https://github.com/typescript-eslint/typescript-eslint/issues/10311
Related, discussion at https://github.com/typescript-eslint/typescript-eslint/pull/10314#discussion_r1836097699
Possibly related, https://github.com/typescript-eslint/typescript-eslint/issues/10415

cc @ronami

### LINKED ISSUE #10443: Bug: [no-unnecessary-boolean-literal-compare] doesn't flag values of a type parameter with boolean type constraint
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=PTAEDMEMEsBsGcBQBjA9gO3gF1FgptgIwBcoAFPKgLZ4DCGAJtFtBqQEaqqx6ToCUoALwA%2BUADdU0BsNCUa9dExYZhYgN6JQoaOHLy6jZq3TChQ3ACcArnkHqAvogcBuRIhByAFqmuwZUHAoGNi4BFgATKQAPAAqoHgAHvhK8KCc3LzoIhTUhkrGbKCxgqISUjIWBorKJmqgmtq6%2Bnk1habmFlg2dg1OrohAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ARhw5IhTWo3yJoQ%2BLTIcAtsRGJ0UJdA7RI4MAF8QZoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// fails
const test1: (someCondition: boolean) => void = someCondition => {
  if (someCondition === true) {}
};

// should fail
const test2: <T extends boolean>(someCondition: T) => void = someCondition => {
  if (someCondition === true) {}
};
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect all use-cases to be consistent and report as errors.

### Actual Result

The use-cases with a type constraint doesn't report as an error (even though it should).

### Additional Info

This seems similar to #10311, and it looks like the fix should be similar.

Note that this issue is also similar to #10442, and I wasn't sure if I should create one issue for each rule or aggregate them under a single issue. I apologize if this was a bit spammy. I checked existing rules for similar problems after resolving #10311.

### PR DESCRIPTION
**Pull request by kirkwaiblinger**
## PR Checklist

- [x] Addresses an existing open issue
   - fixes #10438
   - <s>fixes (not anymore) #10443</s> EDIT - this was already solved in the meantime, just refactoring
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Since removing or modifying `getConstrainedTypeAtLocation` would be a breaking change (see https://typescript-eslint.io/users/versioning#parser-typescript-estree-scope-manager-types-type-utils-utils), <s>I've deprecated it and added a new utility in its place</s> I've left it in place for now, and just added a new utility. Since we may change our implementation in the future, I've made the new utility internal, and not deprecated the existing `getConstrainedTypeAtLocation`. Deprecating it and replacing internal usages can be handled in followup work. See https://github.com/typescript-eslint/typescript-eslint/pull/10496#discussion_r1898269806

I've redone the testing so that the tests for `getConstrainedTypeAtLocation` actually test something useful, and tested several edge cases for both the old and new APIs.

I've used the new API for a few rules for good measure. We'll want to gradually replace the rest of our usages in followup work.

Retouched the fix in #10314 <s>and used the new strategy to solve #10443</s> and in #10474

***
<s>TODO - I'll have to find a way to manage the deprecation linting failures 🙁  https://github.com/typescript-eslint/typescript-eslint/issues/9899 would be extremely clutch. Not really sure how to handle this otherwise, other than to alias the function with a non-deprecated version to replace the existing usages, with the intention of cleaning up the alias later. </s> Not deprecating anything yet.

<s>UPDATE - I chose to alias `getConstrainedTypeAtLocation` as the ironically not-deprecated `DEPRECATED_getConstrainedTypeAtLocation` in order to allow existing usages not to fail CI.</s>

### CODE REVIEW

File: packages/type-utils/tests/getConstrainedTypeAtLocation.test.ts [hunk 1]
```diff
@@ -1,57 +1,309 @@
...
+      mockServices(typeChecker, parameterNode),
+      mockEstreeNode(),
+    );
+
+    expect(tsutils.isTypeParameter(constraintAtLocation)).toBe(false);
+    expect(tsutils.isIntrinsicStringType(constraintAtLocation)).toBe(true);
+  });
+});
+
+describe('getConstraintTypeInfoAtLocation', () => {
```
Comment (kirkwaiblinger) [inline packages/type-utils/tests/getConstrainedTypeAtLocation.test.ts:183]: Add tests for generic whose constraint is generic

File: packages/type-utils/src/getConstrainedTypeAtLocation.ts [hunk 2]
```diff
@@ -4,8 +4,18 @@ import type {
 } from '@typescript-eslint/typescript-estree';
 import type * as ts from 'typescript';
 
+import * as tsApiUtils from 'ts-api-utils';
```
Comment (JoshuaKGoldberg) [inline packages/type-utils/src/getConstrainedTypeAtLocation.ts:7]: [Style] Since this is its name everywhere else:

```suggestion
import * as tsutils from 'ts-api-utils';
```

File: packages/type-utils/package.json [hunk 3]
```diff
@@ -58,6 +58,7 @@
   "devDependencies": {
     "@jest/types": "29.6.3",
     "@typescript-eslint/parser": "8.18.0",
+    "@typescript/vfs": "*",
```
Comment (JoshuaKGoldberg) [inline packages/type-utils/package.json:61]: y?

```suggestion
```

File: packages/type-utils/src/getConstrainedTypeAtLocation.ts [hunk 4]
```diff
@@ -4,8 +4,18 @@ import type {
...
 /**
  * Resolves the given node's type. Will resolve to the type's generic constraint, if it has one.
+ *
+ * If the type is generic and does _not_ have a constraint, the type will be
+ * returned as-is, rather than returning an `unknown` type. This can be checked
+ * for by checking for the type flag ts.TypeFlags.TypeParameter.
+ *
+ * @see https://github.com/typescript-eslint/typescript-eslint/issues/10438
+ *
+ * @deprecated Use getConstraintTypeInfoAtLocation instead.
```
Comment (JoshuaKGoldberg) [inline packages/type-utils/src/getConstrainedTypeAtLocation.ts:18]: [Docs] Fancy linkage:

```suggestion
 * @deprecated Use {@link getConstraintTypeInfoAtLocation} instead.
```

File: packages/type-utils/src/getConstrainedTypeAtLocation.ts [hunk 5]
```diff
@@ -18,3 +28,59 @@ export function getConstrainedTypeAtLocation(
 
   return constrained ?? nodeType;
 }
+
+export interface ConstraintTypeInfoBase {
+  constraintType: ts.Type | undefined;
+  isTypeParameter: boolean;
+  type: ts.Type;
+}
+
```
Comment (JoshuaKGoldberg) [inline packages/type-utils/src/getConstrainedTypeAtLocation.ts:37]: [Style] This doesn't enforce much and isn't used anywhere outside the `extends`:

```suggestion
```

**Review by JoshuaKGoldberg — APPROVED** [review]
LGTM, nice to see the cleanup & related fixes!

Just requesting a collection of touchups, nothing major.

File: packages/type-utils/src/getConstrainedTypeAtLocation.ts (hunk 5 above)
Comment (kirkwaiblinger) [reply packages/type-utils/src/getConstrainedTypeAtLocation.ts:37]: fair enough. i thought this was sort of a required part of the idiom but yeah it can be left off 👍

File: packages/type-utils/package.json (hunk 3 above)
Comment (kirkwaiblinger) [reply packages/type-utils/package.json:61]: 🤷‍♂️ 

https://github.com/typescript-eslint/typescript-eslint/blob/bc12e04be86f7c7dd42d4f2a02f1cf51f35bf067/packages/type-utils/tests/getConstrainedTypeAtLocation.test.ts#L46-L76

Comment (kirkwaiblinger) [reply packages/type-utils/package.json:61]: ❤️

Comment (kirkwaiblinger) [reply packages/type-utils/package.json:61]: Actually, I may be able to copy testing patterns from within this repo, and avoid that, lemme poke around.

**Review by JoshuaKGoldberg — APPROVED** [review]
I like this strategy 😄

File: packages/type-utils/package.json (hunk 3 above)
Comment (kirkwaiblinger) [reply packages/type-utils/package.json]: Yeah, removed all that vfs stuff. Turns out, we already infra for parsing and using type information. _Who would have guessed?_ 🤦‍♂️

File: packages/type-utils/src/getConstrainedTypeAtLocation.ts [hunk 6]
```diff
@@ -18,3 +28,52 @@ export function getConstrainedTypeAtLocation(
...
+  | ConstraintTypeInfoConstrained
+  | ConstraintTypeInfoNonGeneric
+  | ConstraintTypeInfoUnconstrained;
+
+/**
+ * Resolves the given node's type, and returns info about whether it is a generic or ordinary type.
+ *
+ * Successor to {@link getConstrainedTypeAtLocation} due to https://github.com/typescript-eslint/typescript-eslint/issues/10438
+ */
+export function getConstraintTypeInfoAtLocation(
```
Comment (kirkwaiblinger) [inline packages/type-utils/src/getConstrainedTypeAtLocation.ts:60]: @JoshuaKGoldberg I've just had the thought - I think this might be the wrong API to write. It might be better to just accept a type as input, like
```ts
export declare function getConstraintInfo(type: ts.Type): { 
  isTypeParameter: boolean, 
  constraint: ts.Type | undefined
};
```
This maintains a separation between _getting_ types and _analyzing_ types.

Comment (JoshuaKGoldberg) [reply packages/type-utils/src/getConstrainedTypeAtLocation.ts:60]: @kirkwaiblinger just confirming, are you planning on making changes? And/or, is this a question?

Comment (kirkwaiblinger) [reply packages/type-utils/src/getConstrainedTypeAtLocation.ts:60]: Lol. In my head, I was asking you for your opinion, yeah, but I forgot to write that out 😆 

Coming back to this, I currently lean towards: let's go with this PR in its current state, and if there's a need to extract just the constraint info part of this function that's quite easy to do in a followup when the need arises.

Feel free to offer your opinion! 😁

Comment (kirkwaiblinger) [reply packages/type-utils/src/getConstrainedTypeAtLocation.ts:60]: If you're on board with this plan, then this PR is up to date and ready to be merged 👍

Comment (JoshuaKGoldberg) [reply packages/type-utils/src/getConstrainedTypeAtLocation.ts:60]: Oh awesome! I don't have a strong preference here.

Proposal: whatever strategy you want to go with, how about marking the final creation as `@internal` so it's not part of the public API? That way we can play with it as we want.

Comment (kirkwaiblinger) [reply packages/type-utils/src/getConstrainedTypeAtLocation.ts]: Made lots of changes in this arena. New utility is just internal to eslint-plugin for now, no deprecation yet. Should be much easier to review!

**Review by JoshuaKGoldberg — APPROVED** [review]
🎉 This is great!

File: packages/type-utils/src/getConstrainedTypeAtLocation.ts (hunk 6 above)
Comment (JoshuaKGoldberg) [reply packages/type-utils/src/getConstrainedTypeAtLocation.ts]: I like this strategy. The union of possible constraint info points seems reasonable to me.
